src/transport/MessageEndpoint.cpp

```diff
@@ -0,0 +1,302 @@
+            try {
+                this->socket = std::make_unique<zmq::socket_t>(
+                  context.get(), zmq::socket_type::push);
+            } catch (zmq::error_t& e) {
+                logger->error(
+                  "Error opening SEND socket to {}: {}", address, e.what());
+                throw;
+            }
+            break;
+        case faabric::transport::SocketType::PULL:
+            try {
+                this->socket = std::make_unique<zmq::socket_t>(
+                  context.get(), zmq::socket_type::pull);
+            } catch (zmq::error_t& e) {
+                logger->error("Error opening RECV socket bound to {}: {}",
+                              address,
+                              e.what());
+                throw;
+            }
+            break;
+        default:
+            throw std::runtime_error("Unrecognized socket type");
+    }
+    assert(this->socket != nullptr);
+
+    // Bind or connect the socket
+    if (bind) {
+        try {
+            this->socket->bind(address);
+        } catch (zmq::error_t& e) {
+            logger->error("Error binding socket to {}: {}", address, e.what());
+            throw;
+        }
+    } else {
+        try {
+            this->socket->connect(address);
+        } catch (zmq::error_t& e) {
+            logger->error(
+              "Error connecting socket to {}: {}", address, e.what());
+            throw;
+        }
+    }
+}
+
+void MessageEndpoint::send(uint8_t* serialisedMsg, size_t msgSize, bool more)
+{
+    assert(tid == std::this_thread::get_id());
+    assert(this->socket != nullptr);
+
+    const auto& logger = faabric::util::getLogger();
+
+    if (more) {
+        try {
+            auto res = this->socket->send(zmq::buffer(serialisedMsg, msgSize),
+                                          zmq::send_flags::sndmore);
+            if (res != msgSize) {
+                throw std::runtime_error(
+                  fmt::format("Sent different bytes than expected (sent "
+                              "{}, expected {})",
+                              res.value_or(0),
+                              msgSize));
+            }
+        } catch (zmq::error_t& e) {
+            logger->error("Error sending message: {}", e.what());
+            throw;
+        }
+    } else {
+        try {
+            auto res = this->socket->send(zmq::buffer(serialisedMsg, msgSize),
+                                          zmq::send_flags::none);
+            if (res != msgSize) {
+                throw std::runtime_error(
+                  fmt::format("Sent different bytes than expected (sent "
+                              "{}, expected {})",
+                              res.value_or(0),
+                              msgSize));
+            }
+        } catch (zmq::error_t& e) {
+            logger->error("Error sending message: {}", e.what());
+            throw;
+        }
+    }
+}
+
+// TODO - use optional size parameter
```

Comment:
Why would we want an optional size parameter?

Reply:
I've updated the comment, the optional size parameter is aimed at receiving into a pre-allocated buffer. I left a `TODO` as it's currently not used anywhere, but changed it to an informative comment.